src/NUnitFramework/framework/TestContext.cs

```diff
@@ -476,16 +478,110 @@ public TestAdapter(Test test)
             /// <summary>
             /// The expected result if there is one for the test
             /// </summary>
-            public object? ExpectedResult
+            public object? ExpectedResult => (_test as TestMethod)?.ExpectedResult;
+
+            /// <summary>
+            /// The parent of this test or suite
+            /// </summary>
+            public ITest? Parent => _test.Parent;
+
+            /// <summary>
+            /// Returns all properties in the hierarchy
+            /// </summary>
+            /// <returns></returns>
+            public IDictionary<PropertyHierachyItem, IList> PropertyHierarchy()
+            {
+                var dict = new Dictionary<PropertyHierachyItem, IList>();
+                ITest? test = _test;
+                do
+                {
+                    foreach (var property in test.Properties.Keys)
+                    {
+                        var values = test.Properties[property];
+                        dict.Add(
+                            new PropertyHierachyItem { Name = property, Level = test.Name ?? string.Empty }, values);
+                    }
+                    test = test.Parent;
+                }
+                while (test is object);
+
+                return dict;
+            }
+
+            /// <summary>
+            /// Returns all values of a given property
+            /// </summary>
+            /// <param name="property">Name of property</param>
+            public IEnumerable<object> PropertyValues(string property)
+            {
+                var list = new List<object>();
+                var props = PropertyHierarchy();
+                foreach (var item in props.Keys.Where(o => o.Name == property))
```

Comment:
This is expensive, first creating a dictionary of all properties, then filtering.
Depending on the outcome of my previous question of what the result should be.
It can be cheaper:
* Return value of test.Properties[property] if defined, otherwise return value of `test.Parent.Properties[property]` or null if no parent.

Reply:
Ok, I'll have a  look at that.  I am not that concerned about efficiency here, as I expect this to be called rarely.  But your suggestion makes sense, so perhaps the way to go.

Reply:
A proeprty is implemented as a PropertyBag, and can have duplicate entries for the same key.  So, given that, it might be correct that it adds up all the values of a given key from a hierarchy.  
It is mentioned here:  https://docs.nunit.org/articles/nunit/writing-tests/attributes/property.html?q=property  and stated explicitly here:  https://github.com/nunit/nunit/blob/6d6a424626ffc21cbb636016d6c804efe4bfbb3c/src/NUnitFramework/framework/Internal/PropertyBag.cs#L11

But still,  one could use the same and just build up a larger propertybag by walking the Property hierarchy.